File: frontend/src/services/cart-service.ts
import api from "@/lib/api";
import { API_PATHS } from "@/lib/constants";
import type { ApiResponse } from "@/types/api";
import type { CartItem, CartItemRequest, MergeCartRequest } from "@/types/cart";

/**
 * Gets the user's cart.
 */
export async function getCart(): Promise<ApiResponse<CartItem[]>> {
  const response = await api.get<ApiResponse<CartItem[]>>(API_PATHS.CART.GET);
  return response.data;
}

/**
 * Adds an item to the cart.
 */
export async function addToCart(
  data: CartItemRequest
): Promise<ApiResponse<CartItem>> {
  const response = await api.post<ApiResponse<CartItem>>(
    API_PATHS.CART.ADD_ITEM,
    data
  );
  return response.data;
}

/**
 * Updates a cart item's quantity.
 */
export async function updateCartItem(
  productId: number,
  quantity: number
): Promise<ApiResponse<CartItem>> {
  const response = await api.put<ApiResponse<CartItem>>(
    API_PATHS.CART.UPDATE_ITEM(productId),
    null,
    { params: { quantity } }
  );
  return response.data;
}

/**
 * Removes a cart item.
 */
export async function removeCartItem(
  productId: number
): Promise<ApiResponse<void>> {
  const response = await api.delete<ApiResponse<void>>(
    API_PATHS.CART.REMOVE_ITEM(productId)
  );
  return response.data;
}

/**
 * Merges local cart items with the server cart.
 */
export async function mergeCart(
  data: MergeCartRequest
): Promise<ApiResponse<CartItem[]>> {
  const response = await api.post<ApiResponse<CartItem[]>>(
    API_PATHS.CART.MERGE,
    data.items
  );
  return response.data;
}

/**
 * Clears the user's cart (not directly exposed via API; handled server-side on order creation).
 */
export async function clearCart(): Promise<void> {
  // Cart is cleared by the server when an order is created
}

File: backend/src/main/java/com/headless/ecommerce/service/CartService.java
package com.headless.ecommerce.service;

import com.headless.ecommerce.dto.request.CartItemRequest;
import com.headless.ecommerce.dto.request.OrderCreateRequest;
import com.headless.ecommerce.dto.response.CartItemResponse;

import java.util.List;

/**
 * Cart service interface for shopping cart operations.
 */
public interface CartService {

    /**
     * Gets the user's cart items.
     *
     * @param userId the user ID
     * @return the list of cart item responses
     */
    List<CartItemResponse> getCart(Long userId);

    /**
     * Adds an item to the user's cart.
     *
     * @param userId the user ID
     * @param request the cart item request
     * @return the added cart item response
     */
    CartItemResponse addToCart(Long userId, CartItemRequest request);

    /**
     * Updates the quantity of a cart item.
     *
     * @param userId the user ID
     * @param productId the product ID
     * @param quantity the new quantity
     * @return the updated cart item response
     */
    CartItemResponse updateQuantity(Long userId, Long productId, Integer quantity);

    /**
     * Removes an item from the user's cart.
     *
     * @param userId the user ID
     * @param productId the product ID
     */
    void removeItem(Long userId, Long productId);

    /**
     * Merges local cart items with the user's server cart.
     *
     * @param userId the user ID
     * @param localItems the local cart items to merge
     * @return the merged cart items
     */
    List<CartItemResponse> mergeCart(Long userId, List<CartItemRequest> localItems);

    /**
     * Clears the user's cart.
     *
     * @param userId the user ID
     */
    void clearCart(Long userId);
}
